packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -493,8 +493,8 @@ type BoxedHook =
 #[cfg(feature = "std")]
 #[allow(clippy::redundant_pub_crate)]
 pub(crate) struct Hooks {
-    // TODO: Remove `Option` when const `BTreeMap::new` is stabilized
-    pub(crate) inner: Option<BTreeMap<TypeId, BoxedHook>>,
+    // We use `Vec`, instead of `HashMap` or `BTreeMap` so that the ordering stays consistent
```

Comment:
Maybe we should mention, that we make the ordering consistent, but we don't guarantee, that the most recent added hook will be executed at the end (see below). I think with `BTreeMap` this is more consistent, but with `Vec` we have semi-control over the ordering?

Reply:
The problem with `BTreeMap` is that it is ordered by the `TypeId`, which in my experience, can be super annoying from a user perspective.

So instead, we use insertion order.

Reply:
Should we additionally mention this in the `fmt` or `install_debug_hook` docs?

Reply:
I agree, that `BTreeMap` is annoying (but still better than `HashMap`). But also with `Vec` it's not guaranteed to be at the end after inserting because in the case it was found in the existing hooks it get replaced. So if you insert a hook for type `X` first, `X` will always be the first. IMO this is fine and I wouldn't document it. We can treat is as currently unspecified and revisit this later. One solution would be to uninstall a hook and then add the new hook if it already did exist.

Reply:
clarified in [d388a1b](https://github.com/hashintel/hash/pull/974/commits/d388a1b9f48f3acad5d61e3a7afd6dabf371897e)

Reply:
In theory, we could remove the `Vec<>` (I purposefully tried not to do that, as it takes $O(n)$ worst case)

My other idea would've been.

```rust
struct Hooks {
	counter: usize,
	lookup: HashMap<TypeId, usize>,
	inner: BTreeMap<usize, (TypeId, Hook)>
}

impl Hooks {
	fn insert(t: &T) {
		let tid = TypeId::of::<T>();

		let idx = counter += 1;
		
		let prev = self.lookup.insert(&tid);
		if let Some(prev) = prev {
			self.inner.remove(prev);
		}

		self.inner.insert(idx, <boxed hook>);
	}
}
```

This would enable efficient lookups and memory usage but has some more complexity.

Complexity would be $O(1) + O(1) + O(log(n)) \iff O(log(n))$ instead.

Reply:
I don't think it's worth the effort. Sure, an insertion complexity of $O(log(n))$ would be nice, but I think $O(n)$ is fine for now as $n$ should be pretty small.

Reply:
fixed in [b303952](https://github.com/hashintel/hash/pull/974/commits/b303952afebd3291c6290ca50d77d2047854c773), using $O(n)$ approach